domain/model/service/service_test.go

```diff
@@ -1112,3 +1122,74 @@ func (s *serviceSuite) TestDefaultModelCloudNameAndCredential(c *gc.C) {
 		Name:  "test-cred",
 	})
 }
+
+// TestWatchActivatedModels verifies that WatchActivatedModels correctly sets up a watcher
+// that emits events for activated models when the watcher receives change events.
+func (s *serviceSuite) TestWatchActivatedModels(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	ctx := context.Background()
+	svc := NewWatchableService(
+		s.mockState,
+		s.mockModelDeleter,
+		loggertesting.WrapCheckLog(c),
+		s.mockWatcherFactory,
+	)
+
+	activatedModelUUID1 := modeltesting.GenModelUUID(c)
+	activatedModelUUID2 := modeltesting.GenModelUUID(c)
+	activatedModelUUIDs := []coremodel.UUID{activatedModelUUID1, activatedModelUUID2}
+	activatedModelUUIDStrs := transform.Slice(activatedModelUUIDs, func(uuid coremodel.UUID) string {
+		return uuid.String()
+	})
+
+	// Set up necessary state function return values.
+	s.mockState.EXPECT().GetAllActivatedModelsUUIDQuery().Return(
+		"SELECT uuid from model WHERE activated = true",
+	)
+	s.mockState.EXPECT().GetActivatedModelUUIDs(gomock.Any(), gomock.Any()).Return(
+		activatedModelUUIDs, nil,
+	)
+
+	// Create change events for watcher mapper.
+	chEvent1 := changestreammock.NewMockChangeEvent(s.mockWatcherFactory.ctrl)
+	chEvent1.EXPECT().Changed().Return(
+		activatedModelUUID1.String(),
+	)
+
+	chEvent2 := changestreammock.NewMockChangeEvent(s.mockWatcherFactory.ctrl)
+	chEvent2.EXPECT().Changed().Return(
+		activatedModelUUID2.String(),
+	)
+
+	chEvent3 := changestreammock.NewMockChangeEvent(s.mockWatcherFactory.ctrl)
+	chEvent3.EXPECT().Changed().Return(
+		"unactivated model UUID",
+	)
+
+	// Change events received by the watcher mapper.
+	receivedChangeEvents := []changestream.ChangeEvent{chEvent1, chEvent2, chEvent3}
+	// Change events containing model UUIDs of activated models retrieved from the database.
+	expectedChangeEvents := []changestream.ChangeEvent{chEvent1, chEvent2}
+
+	changes := make(chan []string, 1)
+	changes <- activatedModelUUIDStrs
```

Comment:
This is wrong. You are sending a filter result as the raw source events.
You should send the `[]changestream.ChangeEvent{chEvent1, chEvent2, chEvent3}` instead.
Then test the chEvent3 should be excluded by the mapper.

Reply:
Thanks for spotting this Kelvin, will be testing the mapper separately as well :)